main/main.c

```diff
@@ -2199,6 +2199,8 @@ zend_result php_module_startup(sapi_module_struct *sf, zend_module_entry *additi
 	php_ini_register_extensions();
 	zend_startup_modules();
 
+	zend_mm_observers_startup(NULL);
```

Comment:
`zend_mm_observers_startup` would already be called here (`php_module_startup` -> `zend_startup` -> `start_memory_manager` -> `alloc_globals_ctor` -> `zend_mm_observers_startup`). We should be careful not to register the same observers again.

Reply:
`zend_startup()` is run before extension are being loaded, so the observing extension has not been around to register observers yet.
But you catched a nice problem anyway, if the `zend_mm_observers_startup()` function is being called on a heap that already has observers it would leak all allocations from the earlier run.

Reply:
I added the startup path via `alloc_globals_ctor` for ZTS SAPI's (Apache MPM Worker for example) to register the observers with the heap when it creates a new thread.